### ISSUE TITLE
Support functions in PTX

### ISSUE THREAD

**Issue by atzannes** [ufora/ufora#223]
Don’t fully inline all the code. Right now we create a single huge function by agressively inlining every child function that we touch. This has the problem that you can’t do recursion, which PTX now handles. Also, we’ll recompile lots of code each time. PTX allows you to define “functions”
 * Along this line, define the function for “log” once, as opposed to doing it every single time we do a new compilation, which we’re doing now

**Comment by braxtonmckee** [comment]
So it doesn't look like we can compile a block of PTX and then refer to those functions in another block of PTX later. Modules are self-contained in this regard. It is possible to use a linker api (see the ptxjit CUDA example [here](https://github.com/huoyao/cudasdk/blob/master/6_Advanced/ptxjit/ptxjit.cpp)) but it appears that this still requires you to load all the PTX files for a given module at once. So there's not really any point to it for us, because any group of functions that want to talk to each other will still need to be defined all at once, and we might as well just generate a single block of PTX code for it.

**Comment by atzannes** [comment]
Ah, very good to know. Having support for functions would still give us recursion though right?

**Comment by braxtonmckee** [comment]
Yes - but it looks like we have to define the entire suite of functions
together in a single kernel.

On Wed, Mar 9, 2016 at 5:20 AM, Alexandros Tzannes <notifications@github.com
> wrote:

> Ah, very good to know. Having support for functions would still give us
> recursion though right?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ufora/ufora/issues/223#issuecomment-194225812>.
>



-- 
-------------------------------------------------------------------------------------------
Braxton McKee                             CEO / Ufora
www.twitter.com/braxtonmckee       www.ufora.com
